certloader/spiffe_tls_config.go

```diff
@@ -0,0 +1,148 @@
+/*-
+ * Copyright 2019 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package certloader
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+
+	"github.com/spiffe/go-spiffe/spiffe"
+)
+
+type spiffeTLSConfigSource struct {
+	peer *spiffe.TLSPeer
+	log  Logger
+}
+
+type spiffeLogger struct {
+	log Logger
+}
+
+func (l spiffeLogger) Debugf(format string, args ...interface{}) {
+	l.log.Printf("(spiffe) [DEBUG]: "+format, args...)
+}
+
+func (l spiffeLogger) Infof(format string, args ...interface{}) {
+	l.log.Printf("(spiffe) [INFO]: "+format, args...)
+}
+
+func (l spiffeLogger) Warnf(format string, args ...interface{}) {
+	l.log.Printf("(spiffe) [WARN]: "+format, args...)
+}
+
+func (l spiffeLogger) Errorf(format string, args ...interface{}) {
+	l.log.Printf("(spiffe) [ERROR]: "+format, args...)
+}
+
+func TLSConfigSourceFromWorkloadAPI(addr string, log Logger) (TLSConfigSource, error) {
+	peer, err := spiffe.NewTLSPeer(
+		spiffe.WithWorkloadAPIAddr(addr),
+		spiffe.WithLogger(spiffeLogger{log: log}),
+	)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: provide a way to close the peer on graceful shutdown
+	return &spiffeTLSConfigSource{
+		peer: peer,
+		log:  log,
+	}, nil
+}
+
+func (s *spiffeTLSConfigSource) Reload() error {
+	// The config returned by the workload TLSConfig maintains itself. Nothing
+	// to do here.
+	return nil
+}
+
+func (s *spiffeTLSConfigSource) CanServe() bool {
+	return true
+}
+
+func (s *spiffeTLSConfigSource) GetClientConfig(base *tls.Config) (TLSClientConfig, error) {
+	return s.newConfig(base)
+}
+
+func (s *spiffeTLSConfigSource) GetServerConfig(base *tls.Config) (TLSServerConfig, error) {
+	return s.newConfig(base)
+}
+
+func (s *spiffeTLSConfigSource) newConfig(base *tls.Config) (*spiffeTLSConfig, error) {
+	s.log.Printf("waiting for initial SPIFFE Workload API update...")
+	if err := s.peer.WaitUntilReady(context.TODO()); err != nil {
+		return nil, err
+	}
+	s.log.Printf("received SPIFFE Workload API update.")
+
+	return &spiffeTLSConfig{
+		base: base,
+		peer: s.peer,
+	}, nil
+}
+
+type spiffeTLSConfig struct {
+	base *tls.Config
+	peer *spiffe.TLSPeer
+}
+
+func (c *spiffeTLSConfig) GetClientConfig() *tls.Config {
+	config := c.base.Clone()
+	config.InsecureSkipVerify = true
```

Comment:
Add a comment explaining why this is safe?

Reply:
👍